libkbfs/dirty_file.go

```diff
@@ -0,0 +1,161 @@
+// Copyright 2016 Keybase Inc. All rights reserved.
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+package libkbfs
+
+import (
+	"fmt"
+	"sync"
+)
+
+// syncBlockState represents that state of a block with respect to
+// whether it's dirty or currently being synced.  There can be six states:
+//  1) Not dirty or being synced
+//  2) Dirty and awaiting a sync
+//  3) Being synced and not yet re-dirtied: any write needs to
+//     make a copy of the block before dirtying it.  Also, all writes must be
+//     deferred.
+//  4) Being synced and already re-dirtied: no copies are needed, but all
+//     writes must still be deferred.
+//  5) Finished syncing, and not yet re-dirtied, but the rest of the
+//     sync hasn't finished yet.
+//  5) Finished syncing, and already re-dirtied, but the rest of the
+//     sync hasn't finished yet.
+type syncBlockState int
+
+const (
+	blockNotDirty syncBlockState = iota
+	blockDirty
+	blockSyncingNotDirty
+	blockSyncingAndDirty
+	blockSyncedAndDirty
+	blockSyncedNotDirty
+)
+
+// dirtyFile represents a particular file that's been written to, but
+// has not yet completed syncing its dirty blocks to the server.
+type dirtyFile struct {
+	lock sync.Mutex
+	// Which blocks are currently being synced, so that writes and
+	// truncates can do copy-on-write to avoid messing up the ongoing
+	// sync.  If it is blockSyncingNotDirty, then any write to the
+	// block should result in a deep copy and those writes should be
+	// deferred; if it is blockSyncingAndDirty, then just defer the
+	// writes.
+	fileBlockStates map[BlockPointer]syncBlockState
+}
+
+func newDirtyFile() *dirtyFile {
+	return &dirtyFile{
+		fileBlockStates: make(map[BlockPointer]syncBlockState),
+	}
+}
+
+func (df *dirtyFile) blockNeedsCopy(ptr BlockPointer) bool {
+	df.lock.Lock()
+	defer df.lock.Unlock()
+	return df.fileBlockStates[ptr] == blockSyncingNotDirty
+}
+
+// dirtyBlock transitions a block to a dirty state, and returns
+// whether or not the block needs to be put in the dirty cache
+// (because it isn't yet), and whether or not the block is currently
+// part of a sync in progress.
+func (df *dirtyFile) dirtiedBlock(ptr BlockPointer) (
```

Comment:
comment and method have a different name. I'd suggest something to make clear that this is a mutator: `makeBlockDirty` or `setBlockDirty`